<a href="https://colab.research.google.com/github/chiemahp/Flyrank-internship-ml/blob/main/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/chiemahp/Flyrank-internship-ml/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

**Ranking signal analysis:** Identify which page-level engagement and visibility signals are most predictive of decline, and measure their effect sizes and variance. This lane fits because understanding signal strength is foundational to any ranking or classification task, the output drives feature selection for downstream models, and the analysis is interpretable and observable.

In [ ]:
from pathlib import Path
import pandas as pd

# From work/notebooks, go up two levels to reach repo root
repo_root = Path.cwd().parent.parent
data_path = repo_root / "data" / "raw" / "content_refresh_anonymized.csv"
df = pd.read_csv(data_path)

summary = {
    "rows": len(df),
    "clients": df["client_id"].nunique(),
    "pages_per_client": round(len(df) / df["client_id"].nunique(), 1),
    "content_types": df["content_type"].value_counts().head(3).to_dict(),
}

print(summary)

{'rows': 30000, 'clients': 32, 'pages_per_client': 937.5, 'content_types': {'keyword article': 27207, 'feedly article': 2096, 'comparison article': 697}}


## 2. The question: decision, action, cost of a wrong call

**Decision:** Which signals should we prioritize for monitoring or feature engineering?  
**Action:** Engineers and analysts use this to decide which metrics to track, which to include in ranking models, and where to invest instrumentation effort.  
**Cost of a wrong call:** Ignoring a strong signal wastes potential lift; including noise signals wastes compute and obscures true patterns. This is signal analysis, not a causal claim — we're measuring correlation and variance, not intervention.

In [ ]:
from pathlib import Path
import pandas as pd

repo_root = Path.cwd().parent.parent
data_path = repo_root / "data" / "raw" / "content_refresh_anonymized.csv"
df = pd.read_csv(data_path)

signal_cols = ["days_with_impressions", "avg_position", "ctr"]
print(df[signal_cols].describe().round(2).to_string())

       days_with_impressions  avg_position       ctr
count               30000.00      30000.00  30000.00
mean                   61.45         16.34      0.51
std                    32.69         15.22      3.28
min                     1.00          0.00      0.00
25%                    31.00          6.20      0.00
50%                    81.00         10.80      0.07
75%                    88.00         22.30      0.29
max                    88.00        245.00    100.00


## 3. Quick look at the data (2-3 real numbers)

The signals show real variance: position ranges 0–245, days with impressions 1–88, CTR 0–100. About 54% of pages are declining. This spread suggests signals have discriminatory power — some pages rank high on certain metrics while declining, others do not. A signal audit can quantify which signals separate declining from stable pages most strongly.

In [ ]:
from pathlib import Path
import pandas as pd

repo_root = Path.cwd().parent.parent
data_path = repo_root / "data" / "raw" / "content_refresh_anonymized.csv"
df = pd.read_csv(data_path)

df["is_declining_label"] = df["trend_direction"].str.lower().eq("down").astype(int)

print(f"Rows: {len(df):,}")
print(f"Declining-label rate: {df['is_declining_label'].mean():.3f}")
print(f"Average days with impressions: {df['days_with_impressions'].mean():.2f}")
print(f"Average avg_position: {df['avg_position'].mean():.2f}")

Rows: 30,000
Declining-label rate: 0.542
Average days with impressions: 61.45
Average avg_position: 16.34


## 4. Careful words: what I can and can't claim

**I can claim:**  
- Observed: Which signals correlate with declining pages, and by how much (effect sizes, variance, means).
- Measured: Grouped comparisons (declining vs. stable pages) with honest base rates and distributions.
- Directional: Signal A is a stronger predictor than signal B on this dataset.

**I cannot claim:**  
- Causal: Changing a signal causes decline (or prevents it).
- Predictive beyond holdout: Generalization to future unseen data without proper time-aware validation.
- Universal: "This signal always matters" without context on content type, client, or time period.

In [ ]:
from pathlib import Path
import pandas as pd

repo_root = Path.cwd().parent.parent
data_path = repo_root / "data" / "raw" / "content_refresh_anonymized.csv"
df = pd.read_csv(data_path)

df["is_declining_label"] = df["trend_direction"].str.lower().eq("down").astype(int)
base_rate = df["is_declining_label"].mean()

print(f"Observed base rate: {base_rate:.3f}")
print("This supports a decision-support ranking claim, not a causal claim.")

Observed base rate: 0.542
This supports a decision-support ranking claim, not a causal claim.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.